server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1045,9 +1043,8 @@ public static Engine.Index prepareIndex(
             // whether mappings were provided or not.
             doc.addDynamicMappingsUpdate(mapping);
         }
-        Term uid = new Term(IdFieldMapper.NAME, Uid.encodeId(doc.id()));
         return new Engine.Index(
-            uid,
+            Uid.encodeId(doc.id()),
```

Comment:
There's an obvious follow-up here to stop keeping around `String` id on `doc` (or even creating it in the first place) that would simplify this code and save loads of additional allocations in various places.